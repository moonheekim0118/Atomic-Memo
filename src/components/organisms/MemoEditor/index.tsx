import React, { useCallback, useState } from 'react';
import TitleText from '../../molecules/TitleText';
import MainText from '../../molecules/MainText';
import Select from '../../molecules/Select';
import useValidation from '../../../hooks/useValidation';
import { SentMemoData } from '../../../model/memoData';
import { Redirect } from 'react-router-dom';
import { useDispatch } from 'react-redux';
import { createAction, updateAction } from '../../../actions/memo';
import Button from '../../atoms/Button';
import styled from 'styled-components';

interface Props {
  exData?: SentMemoData;
}

const DEFAULT_TEXT = 'Select Kind of Memo!';

const MemoEditor = ({ exData }: Props) => {
  const [mainText, onChangeMainText, mainTextError] = useValidation({
    initialValue: exData ? exData.main : '',
    max: 500,
    min: 0,
  });
  const [titleText, onChangeTitleText, mainTitleError] = useValidation({
    initialValue: exData ? exData.title : '',
    max: 40,
    min: 1,
  });
  const [memoKind, setMemoKind] = useState<string>(
    exData ? exData.kind : DEFAULT_TEXT
  );
  const [redirect, setRedirect] = useState<boolean>(false);

  const dispatch = useDispatch();

  const onSave = useCallback(() => {
    const kind = memoKind === DEFAULT_TEXT ? '🏄🏽‍♀️Life' : memoKind;
    const seconds = new Date().getTime() / 1000; // 현재 시간 구하기
    const data = {
      title: titleText,
      main: mainText,
      kind,
      time: { seconds },
    };
    let actionData;
    if (exData) {
      // edit mode
      data['id'] = exData.id;
      actionData = updateAction(data);
    } else {
      actionData = createAction(data);
    }
    dispatch(actionData);
    setRedirect(true);
  }, [mainText, titleText, memoKind]);

  if (redirect) {
    return <Redirect to="/" />;
  }
  return (
    <StyledMemoEditor>
      <TitleText
        value={titleText}
        onChange={onChangeTitleText}
        error={mainTitleError}
      />
      <MainText
        value={mainText}
        onChange={onChangeMainText}
        error={mainTextError}
      />
      <SelectContainer>
        <Select defaultText={memoKind} setDefaultText={setMemoKind} />
      </SelectContainer>
      <Button color="#3399ff" onClick={onSave}>
        save
      </Button>
    </StyledMemoEditor>
  );
};

const StyledMemoEditor = styled.section`
  display: flex;
  flex-direction: column;
  padding: 2em 0.5em;
  position: relative;
`;

const SelectContainer = styled.div`
  position: absolute;
  top: 112px;
  right: 25px;
`;

export default MemoEditor;
