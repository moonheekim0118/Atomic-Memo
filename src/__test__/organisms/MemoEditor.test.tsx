import 'jsdom-global/register';
import React from 'react';
import MemoEditor from '../../components/organisms/MemoEditor';
import { mount } from 'enzyme';

jest.mock('react-redux', () => ({
  useDispatch: jest.fn(),
}));

describe('<MemoEditor/>', () => {
  let container = null;

  it('renders correctly', () => {
    container = mount(<MemoEditor />);
  });

  it('matches snpashot', () => {
    expect(container.html()).toMatchSnapshot();
  });
});
