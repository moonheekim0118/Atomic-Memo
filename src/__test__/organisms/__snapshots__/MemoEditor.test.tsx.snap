// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MemoEditor/> matches snpashot 1`] = `"<section class=\\"MemoEditor__StyledMemoEditor-sc-1v94iyx-0 iDZeE\\"><div class=\\"TitleText__StyledTitleText-sc-5531il-0 bvunhV\\"><textarea rows=\\"2\\" font-size=\\"2rem\\" placeholder=\\"Title\\" class=\\"Textarea__StyledTextArea-o90l14-0 duiURD\\"></textarea><div class=\\"TitleText__Count-sc-5531il-1 hyqDxi\\"><span color=\\"black\\" width=\\"auto\\" display=\\"inline\\" class=\\"Span__StyledSpan-sc-10y5m6u-0 jaanRG normal\\">0/40</span></div></div><div class=\\"MainText__StyledMainText-sc-17zmbb-0 dgfvXV\\"><textarea rows=\\"10\\" font-size=\\"1.2rem\\" placeholder=\\"Contents\\" class=\\"Textarea__StyledTextArea-o90l14-0 kGrAoo\\"></textarea><span color=\\"black\\" width=\\"auto\\" display=\\"inline\\" class=\\"Span__StyledSpan-sc-10y5m6u-0 jaanRG normal\\">0/500</span></div><div class=\\"MemoEditor__SelectContainer-sc-1v94iyx-1 cVnUje\\"><article class=\\"Select__Container-sc-1dr9cr-0 kPhMeR\\"><div class=\\"Select__DefaultText-sc-1dr9cr-1 jGhHrZ selected-text\\">Select Kind of Memo!</div></article></div><button color=\\"#3399ff\\" class=\\"Button__StyledButton-hxb5gx-0 dfmXPz\\">save</button></section>"`;
